File: src/main/scala/com/wts/kayan/app/job/MainDriver.scala
package com.wts.kayan.app.job

import java.io.File

import com.typesafe.config.{Config, ConfigFactory}
import com.wts.kayan.sessionmanager.SparkSessionManager
import com.wts.kayan.app.common.PrimaryRunner
import com.wts.kayan.app.reader.PrimaryReader
import com.wts.kayan.app.utility.PrimaryConstants
import com.wts.kayan.app.writer.PrimaryWriter
import org.slf4j.LoggerFactory

object MainDriver {

  private val logger = LoggerFactory.getLogger(this.getClass)

  /**
    *
    * @param args jar parameters
    */
  def main(args: Array[String]): Unit = {

    if (args.isEmpty || args(0).trim.isEmpty) {
      logger.error("Missing required argument. Usage: MainDriver <env> [configPath] (e.g. dev localRun/application.conf)")
      throw new IllegalArgumentException(
        "Missing required argument <env>. Usage: MainDriver <env> [configPath] (e.g. dev localRun/application.conf)")
    }

    val env = args(0).trim

    // args(1) (optional): path to an external application.conf (e.g. localRun/application.conf).
    // When provided, it takes precedence over the bundled classpath config; otherwise the
    // classpath application.conf (src/main/resources) is used.
    implicit val config: Config =
      if (args.length > 1 && args(1).trim.nonEmpty) {
        val configFile = new File(args(1).trim)
        if (!configFile.exists()) {
          throw new IllegalArgumentException(
            s"Config file not found: ${configFile.getAbsolutePath} " +
              s"(JVM working directory ${new File(".").getAbsolutePath}). " +
              s"Pass a path relative to the working directory, or an absolute path.")
        }
        logger.info(s"\n**** Loading external configuration from ${configFile.getAbsolutePath} ****\n")
        ConfigFactory.parseFile(configFile).withFallback(ConfigFactory.load()).resolve()
      } else {
        logger.info("\n**** Loading bundled configuration from classpath (application.conf) ****\n")
        ConfigFactory.load()
      }

    // Fail fast with an actionable message rather than a cryptic "No setting for key" later on.
    if (!config.hasPath("training_spark")) {
      throw new IllegalStateException(
        "Loaded configuration has no 'training_spark' section. Either pass an external config as the " +
          "second argument (e.g. 'dev localRun/application.conf'), or rebuild so application.conf is on " +
          "the classpath (in IntelliJ: Build > Rebuild Project).")
    }

    val sparkSession = SparkSessionManager.fetchSparkSession(PrimaryConstants.APPLICATION_NAME)

    logger.info(s"\n\n****  training job has started ... **** \n\n", this.getClass.getName)

    val primaryReader = new PrimaryReader()(sparkSession, env, config)
    val primaryWriter = new PrimaryWriter()(env)
    val primaryRunner = new PrimaryRunner(primaryReader)(sparkSession).runPrimaryRunner()

    primaryWriter.write(primaryRunner, PrimaryConstants.CLIENTS_ORDERS, PrimaryConstants.MODE_OVERWRITE, 2)(env, config)

    // In debug mode, keep the JVM (and Spark UI) alive so it can be browsed before shutdown.
    if (config.getBoolean("training_spark.debug.enabled")) {
      val sleepMillis = config.getLong("training_spark.debug.ui_sleep_millis")
      logger.info(s"\n\n**** debug mode enabled - keeping Spark UI alive for $sleepMillis ms **** \n\n")
      Thread.sleep(sleepMillis)
    }

    sparkSession.stop()
  }
}

File: src/main/java/com/wts/kayan/app/job/MainDriver.java
package com.wts.kayan.app.job;

import com.wts.kayan.app.common.PrimaryRunner;
import com.wts.kayan.app.reader.PrimaryReader;
import com.wts.kayan.app.utility.AppConfig;
import com.wts.kayan.app.utility.PrimaryConstants;
import com.wts.kayan.app.writer.PrimaryWriter;
import com.wts.kayan.sessionmanager.SparkSessionManager;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Entry point — Enhancement branch: HOCON config, Log4j2, unit tests.
 *
 * <p><strong>What changed vs. ud04:</strong>
 * <ol>
 *   <li>{@link AppConfig} replaces the raw {@code env} string as the configuration
 *       carrier — all paths and writer settings come from {@code application.conf}.</li>
 *   <li>Log4j2 is now the SLF4J backend (configured via {@code log4j2.xml}):
 *       Spark/Hadoop logs are silenced to WARN, application logs stay at INFO.</li>
 *   <li>{@link PrimaryWriter#write(Dataset, AppConfig)} takes a single {@code AppConfig}
 *       argument — no more scattered magic numbers.</li>
 * </ol>
 *
 * @author Mehdi TAJMOUATI
 * @see <a href="https://www.wytasoft.com/wytasoft-group/">WyTaSoft — courses and training sessions</a>
 */
public class MainDriver {

    private static final Logger logger = LoggerFactory.getLogger(MainDriver.class);

    public static void main(String[] args) {

        // -------------------------------------------------------------------
        // Step 1 — Load configuration from application.conf (classpath)
        // -------------------------------------------------------------------
        // AppConfig wraps Typesafe ConfigFactory.load() and exposes typed
        // accessors. The env string selects the right path block in the HOCON file.
        String    env       = args[0];
        AppConfig appConfig = new AppConfig(env);

        // -------------------------------------------------------------------
        // Step 2 — Create the SparkSession
        // -------------------------------------------------------------------
        SparkSession sparkSession = SparkSessionManager.fetchSparkSession(
                PrimaryConstants.APPLICATION_NAME
        );

        logger.info("\n\n****  training job has started (env={}) ****\n\n", env);

        // -------------------------------------------------------------------
        // Step 3 — EXTRACT
        // -------------------------------------------------------------------
        PrimaryReader primaryReader = new PrimaryReader(sparkSession, appConfig);

        // -------------------------------------------------------------------
        // Step 4 — TRANSFORM
        // -------------------------------------------------------------------
        PrimaryRunner primaryRunner = new PrimaryRunner(primaryReader, sparkSession);
        Dataset<Row>  enriched      = primaryRunner.runPrimaryRunner();
        enriched.show();

        // -------------------------------------------------------------------
        // Step 5 — LOAD
        // -------------------------------------------------------------------
        // All write settings (path, mode, partitions) come from AppConfig.
        new PrimaryWriter().write(enriched, appConfig);

        // -------------------------------------------------------------------
        // Step 6 — Stop
        // -------------------------------------------------------------------
        sparkSession.stop();
    }
}
